### PR TITLE
New property_by_postcode updater to include new fields necessary for query ordering

### PR DESCRIPTION
**Pull request by camilleldn**
(for review only: do not merge yet, waiting for API code and acceptance tests first)

### CODE REVIEW

**Comment by Pheonnexx** [comment]
Check it all, seemed good to me.

File: service/updaters/property_by_postcode_updater_v3.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        else:
+            return self._prepare_upsert_actions(title)
+
+    def get_mapping(self):
+        return {
+            'properties': {
+                'title_number': {'type': 'string', 'index': 'no'},
+                'postcode': {'type': 'string', 'index': 'not_analyzed'},
+                'street_name': {'type': 'string', 'index': 'no'},
+                'house_no': {'type': 'integer', 'index': 'no'},
```
Comment (lgonczar) [inline service/updaters/property_by_postcode_updater_v3.py]: Is `house_no` always an integer?

File: service/updaters/property_by_postcode_updater_v3.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                'house_alpha': {'type': 'string', 'index': 'no'},
+                'street_name_2': {'type': 'string', 'index': 'no'},
+                'secondary_house_no': {'type': 'integer', 'index': 'no'},
```
Comment (lgonczar) [inline service/updaters/property_by_postcode_updater_v3.py]: There's also `sub_building_no` field. Not sure which one we need.

File: service/updaters/property_by_postcode_updater_v3.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            address = title.register_data['address']
+            first_number = self._first_number_not_in_postcode(address['address_string'])
+            id = self._get_document_id(title.title_number, normalised_postcode)
+            house_no = address.get('house_no', None)
+            if house_no:
+                house_int = int(house_no)
+            else:
+                house_int = None
+            secondary_house_no = address.get('secondary_house_no', None)
+            if house_no:
```
Comment (lgonczar) [inline service/updaters/property_by_postcode_updater_v3.py]: Was probably meant to be `if secondary_house_no`

File: tests/test_property_by_postcode_updater_v3.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+        updater = PropertyByPostcodeUpdaterV3(index_name, doc_type)
+        returned_actions = updater.prepare_elasticsearch_actions(deleted_title)
+
+        mock_get_delete_action.assert_called_once_with(index_name, doc_type, title_id)
+
+        assert returned_actions == [{'delete': 'action1'}]
+
+    @mock.patch('service.es_utils.get_upsert_action', return_value={'upsert': 'action1'})
+    def test_prepare_elasticsearch_actions_returns_upsert_action_when_title_open(
```
Comment (lgonczar) [inline tests/test_property_by_postcode_updater_v3.py]: I guess this test should be deleted, as it's replaced by the test below

**Comment by lgonczar** [comment]
Looks good, left some comments. 

I'm wondering about two things:

- Performance - is the sorting by multiple fields going to be performant? (that's really my lack of knowledge)
- Fields other than `address_string` are often missing. Are the addresses going to be sorted correctly for Nones?

File: service/updaters/property_by_postcode_updater_v3.py (hunk 2 above)
Comment (camilleldn) [inline service/updaters/property_by_postcode_updater_v3.py]: now added, together with sub_building_description

File: service/updaters/property_by_postcode_updater_v3.py (hunk 1 above)
Comment (camilleldn) [inline service/updaters/property_by_postcode_updater_v3.py]: no :sob: now added some extra code to work around this and yet keep integer-style ordering for the house_no

**Comment by camilleldn** [comment]
@lgonczar:
- performance should be fine, i'll show you my benchmarks on Monday when you're back
- because the sorts default to "missing" -> "last", and because we try extracting a non-postcode integer on address strings, this feels like a pretty decent attempt in the case of postcode search

**Comment by camilleldn** [comment]
Jenkins build currently broken (tests fine, Jenkins acting up, reported to Mikael :( )

Please merge with associated pull requests on api (https://github.com/LandRegistry/digital-register-api/pull/57) and feeder (https://github.com/LandRegistry/digital-register-feeder/pull/160) :) 
merge order should be:
1) updater
2) feeder and api
thanks :heart:

File: tests/test_property_by_postcode_updater_v3.py (hunk 4 above)
Comment (camilleldn) [inline tests/test_property_by_postcode_updater_v3.py]: yes :facepalm: removed :)

**Comment by lgonczar** [comment]
Good to merge

**Comment by Pheonnexx** [comment]
LGTM

**Pull request merged**